File: app/server/notify/channels/wxpusher.ts
/**
 * WxPusher 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'wxpusher' });

export const wxpusherChannel: NotifyChannel = {
    name: 'wxpusher',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('WXPUSHER_APP_TOKEN')) {
                resolve({ success: false, message: 'WXPUSHER_APP_TOKEN 未配置' });
                return;
            }

            const WXPUSHER_APP_TOKEN = getConfig('WXPUSHER_APP_TOKEN');
            const WXPUSHER_TOPIC_IDS = getConfig('WXPUSHER_TOPIC_IDS');
            const WXPUSHER_UIDS = getConfig('WXPUSHER_UIDS');

            // 处理topic_ids，将分号分隔的字符串转为数组
            const topicIds = WXPUSHER_TOPIC_IDS
                ? WXPUSHER_TOPIC_IDS.split(';')
                    .map((id) => id.trim())
                    .filter((id) => id)
                    .map((id) => parseInt(id))
                : [];

            // 处理uids，将分号分隔的字符串转为数组
            const uids = WXPUSHER_UIDS
                ? WXPUSHER_UIDS.split(';')
                    .map((uid) => uid.trim())
                    .filter((uid) => uid)
                : [];

            // topic_ids uids 至少有一个
            if (!topicIds.length && !uids.length) {
                resolve({ success: false, message: 'WXPUSHER_TOPIC_IDS 和 WXPUSHER_UIDS 至少设置一个' });
                return;
            }

            const url = 'https://wxpusher.zjiecode.com/api/send/message';
            const body = {
                appToken: WXPUSHER_APP_TOKEN,
                content: `<h1>${text}</h1><br/><div style='white-space: pre-wrap;'>${desp}</div>`,
                summary: text,
                contentType: 2, // HTML格式
                topicIds: topicIds,
                uids: uids,
                verifyPayType: 0,
            };

            $.post({ url, json: body }, (err, _resp, data) => {
                if (err) {
                    logger.error({ err }, 'WxPusher 发送失败');
                    resolve({ success: false, error: err as Error });
                } else {
                    const result = data as { code?: number; msg?: string };
                    if (result?.code === 1000) {
                        logger.info('WxPusher 发送成功');
                        resolve({ success: true, message: '发送成功' });
                    } else {
                        logger.warn({ body: data }, 'WxPusher 发送失败');
                        resolve({ success: false, message: result?.msg || '发送失败' });
                    }
                }
            });
        });
    }
};

if (hasConfig('WXPUSHER_APP_TOKEN')) {
    wxpusherChannel.enabled = true;
}

export default wxpusherChannel;

File: app/server/internal/notify/channels/wxpusher.go
package channels

import (
	"fmt"
	"html"
	"log/slog"
	"strconv"
	"strings"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type WxPusher struct {
	enabled bool
}

func (c *WxPusher) Name() string {
	return "wxpusher"
}

func (c *WxPusher) Enabled() bool {
	return c.enabled
}

func (c *WxPusher) SetEnabled(enabled bool) {
	c.enabled = enabled
}

func (c *WxPusher) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("WXPUSHER_APP_TOKEN") {
		return notify.NotifyResult{Success: false, Message: "WXPUSHER_APP_TOKEN 未配置"}
	}

	wxpusherAppToken := notify.GetConfig("WXPUSHER_APP_TOKEN")
	wxpusherTopicIDs := notify.GetConfig("WXPUSHER_TOPIC_IDS")
	wxpusherUIDs := notify.GetConfig("WXPUSHER_UIDS")

	// Parse topicIds from semicolon-separated string
	var topicIds []int
	if wxpusherTopicIDs != "" {
		parts := strings.Split(wxpusherTopicIDs, ";")
		for _, p := range parts {
			p = strings.TrimSpace(p)
			if p != "" {
				if id, err := strconv.Atoi(p); err == nil {
					topicIds = append(topicIds, id)
				}
			}
		}
	}

	// Parse uids from semicolon-separated string
	var uids []string
	if wxpusherUIDs != "" {
		parts := strings.Split(wxpusherUIDs, ";")
		for _, uid := range parts {
			uid = strings.TrimSpace(uid)
			if uid != "" {
				uids = append(uids, uid)
			}
		}
	}

	// At least one of topicIds or uids must be provided
	if len(topicIds) == 0 && len(uids) == 0 {
		return notify.NotifyResult{Success: false, Message: "WXPUSHER_TOPIC_IDS 和 WXPUSHER_UIDS 至少设置一个"}
	}

	url := "https://wxpusher.zjiecode.com/api/send/message"
	body := map[string]interface{}{
		"appToken":       wxpusherAppToken,
		"content":        fmt.Sprintf("<h1>%s</h1><br/><div style='white-space: pre-wrap;'>%s</div>", html.EscapeString(text), html.EscapeString(desp)),
		"summary":        html.EscapeString(text),
		"contentType":    2, // HTML format
		"topicIds":       topicIds,
		"uids":           uids,
		"verifyPayType":  0,
	}

	resp, err := notify.HTTPPost(url, notify.HttpRequestOptions{
		JSON: body,
	})
	if err != nil {
		slog.Error("WxPusher 发送失败", "error", err)
		return notify.NotifyResult{Success: false, Message: err.Error(), Err: err}
	}

	if resp.StatusCode == 200 {
		slog.Info("WxPusher 发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	slog.Warn("WxPusher 发送失败", "status", resp.StatusCode, "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: "发送失败"}
}

func init() {
	ch := &WxPusher{}
	if notify.HasConfig("WXPUSHER_APP_TOKEN") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
